src/index.tsx

```diff
@@ -39,31 +47,45 @@ export default class ReactDetectOfflineAPI extends React.Component<Props, State>
       apiUrl,
     } = this.props;
 
-    timeout(3000, fetch(apiUrl).then(() => {
-      this.setState(({online}) => {
-        if (this.props.onOnline && online === false) {
-          this.props.onOnline()
-        }
-
-        if (this.props.initialStatusCallback && online === null) {
-          this.props.initialStatusCallback(true)
-        }
-
-        return {online: true}
-      })
-    })).catch(() => {
-      this.setState(({online}) => {
-        if (this.props.onOffline && online === true) {
-          this.props.onOffline()
-        }
-
-        if (this.props.initialStatusCallback && online === null) {
-          this.props.initialStatusCallback(false)
-        }
-
-        return {online: false}
+    timeout(3000, fetch(apiUrl)
+      .then(handleErrors)
+      .then(res => res.json())
```

Comment:
I think it'd be useful if users can actually define how promise should be resolved completely. Now we're forcing them to return JSON. What if they want to return `204 No Content`, or even `XML` e.g. in legacy systems?

Instead of parsing `json` and implementing the arrow function down there (53-57). I'd just pass this promise to a `resolveOnlineStatusFromPromise` function which accepts Promise and returns offline/online.

You can implement default prop for this function as part of OOTB implementation.